controllers/useraccount/useraccount_controller.go

```diff
@@ -155,17 +156,33 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 	return reconcile.Result{}, r.setStatusReady(userAcc)
 }
 
+// (optional) migration step:
+// - remove the OwnerReferences between the UserAccount and its associated NSTemplateSet (part of decoupling/CRT1305)
+func (r *Reconciler) migrateNStemplateSetIfNecessary(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (bool, error) {
+	// create if not found
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name, Namespace: userAcc.Namespace}, nsTmplSet); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("NSTemplateSet not found")
+			return false, nil
+		}
+		return false, err
+	}
+	// ensure that the NSTemplateSet `OwnerReferences` is empty
+	if len(nsTmplSet.OwnerReferences) == 0 {
+		return false, nil
+	}
+	nsTmplSet.OwnerReferences = nil
+	return true, r.Client.Update(context.TODO(), nsTmplSet)
```

Comment:
Could you please add an e2e test-case for this? Just to be sure that we won't break the already existing CRs.

Reply:
It's not a trivial thing to verify because the operators are always running, but I've added this test to verify that if an NSTemplateSet has an OwnerReference, it is (immediately) removed by the member operator (the `UserAccountController`): https://github.com/codeready-toolchain/toolchain-e2e/pull/407/commits/8078f6dadb44c1fd8a00f307ea4db12ce5d1cb71

Let me know what you think about it.

Reply:
Yes, this is exactly it - thanks :+1: